### ISSUE TITLE
Incorrect Intensity Calculation in FullyNN

### ISSUE THREAD

**Issue by mlpotter** [ant-research/EasyTemporalPointProcess#68]
I was simulating and fitting to toy synthetic data (for 1 event type) related to the Hawkes process, and I noticed that the `FullyNN` intensity function values seemed to be many magnitudes lower than the true Hawkes process intensity function values.

After some review, I noticed in line 73 of the `torch_fullynn.py` file that there seemed to be a mistake:

```
    derivative_integral_lambda = grad(
                integral_lambda.sum(dim=-1).mean(),
                time_delta_seqs,
                create_graph=True, retain_graph=True)[0]
                derivative_integral_lambda = derivative_integral_lambda.unsqueeze(-1).expand(*derivative_integral_lambda.shape, self.num_event_types) / self.num_event_types
```

Specifically `integral_lambda.sum(dim=-1).mean()` should just be changed to `integral_lambda.sum(dim=-1).sum()`. If you leave the mean for example, then the intensity function values are divided by the batch size x sequence length. Mean is typically in loss functions, but this is for intensity function.

This also changes the log likelihood values, as the intensity function values change quite dramatically.

**Comment by iLampard** [comment]
Hi, it looks reasonable. Can you post some test result to see if the fix can help to recover the intensities?

**Comment by mlpotter** [comment]
Sure. As a reference, I used `hawkesbook` library to simulate the Hawkes process with exponential kernel (mu=0.2,alpha=0.8, and beta=1.0) with the observation length of a session randomly sampled between Tmin=50 and Tmax=100. I generated 50 sequences for training. 

Then, I used `hawkesbook` computation of the intensity and compared to intensity calculated from `FullyNN` model (before and after fixing the gradient calculation.

**Comment by ajboyd2** [comment]
Good find! This is correct, for `grad` to work properly we need to pass in a scalar value, but taking the mean instead of a sum over the values will in turn scale the resulting gradient. 

The corrected snippet of code should be this:
```python
# [batch_size, seq_len, num_event_types]
if self.proper_marked_intensities:
    derivative_integral_lambdas = []
    for i in range(integral_lambda.shape[-1]):  # iterate over marks
        derivative_integral_lambdas.append(grad(
            integral_lambda[..., i].sum(),
            time_delta_seqs,
            create_graph=True, retain_graph=True)[0])
    derivative_integral_lambda = torch.stack(derivative_integral_lambdas, dim=-1) 
else:
    derivative_integral_lambda = grad(
        integral_lambda.sum(),
        time_delta_seqs,
        create_graph=True, retain_graph=True)[0]
    derivative_integral_lambda = derivative_integral_lambda.unsqueeze(-1).expand(*derivative_integral_lambda.shape, self.num_event_types) / self.num_event_types
```

Note in general, do be sure to set `proper_marked_intensities=True` in the config. Otherwise a poor assumption is made on the resulting marked intensities.
